src/Embedders/Word2Vec.php

```diff
@@ -0,0 +1,924 @@
+    }
+
+    /**
+     * Iterating through a range of the specific epoch count and updating all respective word vectors.
+     *
+     * @param \Rubix\ML\Datasets\Dataset $dataset
+     * @throws \InvalidArgumentException
+     */
+    public function fit(Dataset $dataset) : void
+    {
+        DatasetIsNotEmpty::check($dataset);
+
+        $sentences = $dataset->column(0);
+
+        $this->preprocess($sentences);
+        $this->prepareWeights();
+
+        $startAlpha = $this->alpha;
+
+        for ($i = 0; $i < $this->epochs; ++$i) {
+            $this->alpha = $startAlpha - (($startAlpha - self::MIN_ALPHA) * ($i) / $this->epochs);
+
+            $this->trainEpochSg();
+        }
+
+        $this->generateL2Norm();
+    }
+
+    /**
+     * Embed a high dimensional dataset into a lower dimensional one.
+     *
+     * @param array[] $samples
+     * @throws \RuntimeException
+     */
+    public function transform(array &$samples) : void
+    {
+        if (!$this->vectorsNorm) {
+            throw new RuntimeException('Transformer has not been fitted.');
+        }
+
+        foreach ($samples as &$sample) {
+            foreach ($sample as $i2 => $sentence) {
+                $preppedSentence = $this->prepSentence($sentence);
+
+                $embeddings = [];
+                foreach ($preppedSentence as $word) {
+                    $embeddings[] = $this->embedWord($word);
+                }
+
+                unset($sample[$i2]);
+            }
+
+            if (isset($embeddings)) {
+                $sample = Matrix::stack($embeddings)->transpose()->mean()->asArray();
+            }
+        }
+    }
+
+    /**
+     * Determining the top N similar words provided an array of positive words and negative words.
+     *
+     * @param string[] $positive
+     * @param string[] $negative
+     * @param int $top
+     * @return string[] $result
+     */
+    public function mostSimilar(array $positive, array $negative = [], $top = 20) : array
+    {
+        $positiveArray = $negativeArray = $allWords = $means = [];
+
+        foreach ($positive as $word) {
+            $positiveArray[$word] = 1.0;
+        }
+
+        foreach ($negative as $word) {
+            $negativeArray[$word] = -1.0;
+        }
+
+        $wordArray = array_merge($positiveArray, $negativeArray);
+
+        foreach ($wordArray as $word => $weight) {
+            $wordEmbedding = $this->wordVec($word);
+
+            if ($wordEmbedding instanceof Vector) {
+                $means[] = $wordEmbedding->multiplyScalar($weight);
+                $allWords[] = $this->vocab[$word]['index'];
+            }
+        }
+
+        if (empty($allWords)) {
+            throw new InvalidArgumentException('Positive words were not found in vocab.');
+        }
+
+        $mean = Matrix::stack($means)->transpose()->mean();
+        $l2 = Matrix::stack($this->vectorsNorm);
+        $dists = $mean->transpose()->matmul($l2->transpose())->asArray()[0];
+
+        arsort($dists, 1);
```

Comment:
Can we use the constant here instead of an integer to be more expressive?

https://www.php.net/manual/en/function.sort.php